Include post-release updates to `CHANGELOG.md` in release process (#1109)
mongodb/mongo-cxx-driver: etc/releasing.md

User-facing issues should generally be either "Bug" or "New Feature".
Non-user facing issues should generally be "Task" tickets.

## Update CHANGELOG.md pre-release ...

### ... for a minor release (e.g. `1.2.0`)

Create a new branch to contain changelog updates from the master branch: `git checkout -b pre-release-changes master`. This branch will be used to create a PR.

Check Jira for tickets closed in this fix version. Update CHANGELOG.md with notable changes not already mentioned. Remove `[Unreleased]` from the version being released.

> Check if there is an `[Unreleased]` section for a patch version (e.g. `1.1.4 [Unreleased]`). The C++ driver does not typically do patch releases for previously released minor versions. If the previous patch release is no longer planned, move the patch release entries into the minor release section and remove the patch release section. If you are unsure whether the patch release is planned, ask in the #dbx-c-cxx channel.

Reviewer: ```suggestion
Check if there is an `[Unreleased]` section for a patch version (e.g. `1.1.4 [Unreleased]`). Normally, the C++ Driver does not release patches for old minor versions. If an unreleased patch release section is no longer applicable, move its entries into the minor release section (as needed) and remove the patch release section. If you are unsure whether the patch release is planned, ask in the #dbx-c-cxx channel.
```
Slight wording adjustments to improve clarity.

Example (if `1.2.0` is being released):
```md

> ## 1.2.0 [Unreleased]
> ### Fixed
> - Fixes A <!-- Unreleased fix on master branch -->
> ## 1.1.4 [Unreleased]
> ### Fixed
> - Fixes B <!-- Unreleased fix on `releases/v1.1` branch. B is implicity fixed on 1.2.0. Change was cherry-picked from master. -->

Reviewer: ```suggestion
## 1.2.0 [Unreleased]

### Fixed

- Fixes A <!-- Unreleased fix on master branch -->

## 1.1.4 [Unreleased]

### Fixed

- Fixes B <!-- Unreleased fix on `releases/v1.1` branch. B is implicity fixed on 1.2.0. Change was cherry-picked from master. -->
```
Add whitespacing.

```

Becomes:
```md

> ## 1.2.0
> ### Fixed
> - Fixes A
> - Fixes B <!-- 1.1.4 is no longer planned. Document B in 1.2.0 -->

Reviewer: ```suggestion
## 1.2.0

### Fixed

- Fixes A
- Fixes B <!-- 1.1.4 is no longer planned. Document B in 1.2.0 -->
```
Add whitespacing.

```

> Commit with a message like `Update CHANGELOG for <version>`. Create a PR. Once the PR is merged, proceed with the rest of the releease.

Reviewer: Suggestion retracted.

> Commit with a message like `Update CHANGELOG for <version>`. Create a PR. Once the PR is merged, proceed with the rest of the release. The `pre-release-changes` branch may be deleted.

Reviewer: ```suggestion
Commit with a message like `Update CHANGELOG for <version>`. Create a PR for the `pre-release-changes` branch which contains the commits made up to this point. Once the PR is merged, proceed with the rest of the release. The `pre-release-changes` branch may be deleted.
```
Emphasize the PR is for the `pre-release-changes` branch.

### ... for a patch release (e.g. `1.2.3`)

> Check out the release branch (e.g. `releases/v1.2`).

Reviewer: ```suggestion
Check out the existing release branch (e.g. `releases/v1.2`).
```
Emphasis that it _should_ already exist.

Check Jira for tickets closed in this fix version. Update CHANGELOG.md with notable changes not already mentioned. Remove `[Unreleased]` from the version being released.

> Commit with a message like `Update CHANGELOG for <version>`. Push the change.

Reviewer: Do we want to do the `pre-release-changes` method here as well for consistency (+ accounting for the upcoming stable branch policies)?

git push -f origin releases/stable
```

> ## Branch if necessary

Reviewer: Suggestion retracted.

If doing a new minor release `x.y.0` (e.g. a `1.2.0` release), create branch

> `releases/vx.y`  (e.g `releases/v1.2`).

Reviewer: Suggestion retracted.

Push the new branch:

```
git push --set-upstream origin releases/v1.2
```

The new branch should be continuously tested on Evergreen. Create a BUILD ticket
to request the build team create new Evergreen project to track the
`releases/vx.y` branch (see BUILD-5666 for an example).

## Generate and Publish Documentation

Documentation generation must be run after the release tag has been made and
pushed.

> - Checkout the master branch. Create a new branch to contain documentation updates: `git checkout -b post-release-changes`. This branch will be used to create a PR later.

Reviewer: Suggestion retracted.

git push -f origin releases/stable
```

## Branch if necessary

If doing a new minor release `X.Y.0` (e.g. a `1.2.0` release), create branch
`releases/vX.Y` (e.g `releases/v1.2`): `git checkout -b releases/v1.2 master`

Push the new branch:

```
git push --set-upstream origin releases/v1.2
```

The new branch should be continuously tested on Evergreen. Update the "Display Name" and "Branch Name" of the [mongo-cxx-driver-latest-release Evergreen project](https://spruce.mongodb.com/project/mongo-cxx-driver-latest-release/settings/general) to refer to the new release branch.

## Generate and Publish Documentation

Documentation generation must be run after the release tag has been made and
pushed.

> - Create a new branch to contain documentation updates from master: `git checkout -b post-release-changes master`. This branch will be used to create a PR later.

Reviewer: ```suggestion
- Create and checkout a new branch to contain documentation updates from master: `git checkout -b post-release-changes master`. This branch will be used to create a PR later.
```
Emphasis the new branch should be checked out for the following steps (editing `etc`, `docs`, etc.).

On the other hand, if `post-release-changes` is _not_ the branch we should be on during the edits, add a bullet describing what branch we should checkout instead. Prior instructions suggest we should be on the `master` branch after this step:

```
- Create a new branch ...: `git branch post-release-changes master`. This branch will be used to create a PR later.
- Checkout the `master` branch.
- Edit ...
```

The explicit instruction to checkout the `master` branch should account for the various possible currently-checked-out branches that may result from prior instructions.

Note: `git checkout -b <branchname> <ref>` creates _and_ checks out a new branch `<branchname>` made on `<ref>`. To create but _not_ check out a new branch, `git branch <branchname> <ref>` should be used instead (or simply `git branch <branchname>` if we are already on `<ref>`).

  `docs/`. This may require pushing the commit to a fork of the C++ Driver
  repository and creating a pull request.

## Update CHANGELOG.md for next release

### For a minor release (e.g. r1.2.0):

Open a PR to the master branch to add a new empty `[Unreleased]` section to the CHANGELOG.md for the next minor release. Example:

```md
## 1.3.0 [Unreleased]
```

Open a PR to the release branch (e.g. `releases/v1.2`) to add a new empty `[Unreleased]` section to the CHANGELOG.md for the next patch release. Example:

```md
## 1.2.1 [Unreleased]

> ```

Reviewer: Does this imply the main branch should never contain changelog entries for patch releases since they'd be targeting the next minor (or major) release instead?

e.g. https://github.com/mongodb/mongo-cxx-driver/pull/1107 merged 3.10.1 changelog entries to `master` then cherry-picked them to `releases/v3.10`. Should the changelog in `master` be targeting `3.11.0 [Unreleased]` instead?

Author: > Should the changelog in master be targeting 3.11.0 `[Unreleased]` instead?

Yes. I think the `[Unreleased]` changelog entries should track the next release on the same branch:

- Minor releases are made from the master branch. Track the next minor release in the master branch changelog.
- Patch releases are made from release branches. Track the next patch release in the release branch changelog.

That may ease updating the changelog as part of applying the changes intended for the same release.

I updated the changelog in this PR to include `## 3.11.0 [Unreleased]` and remove `## 3.10.1 [Unreleased]`.

> Does this imply the main branch should never contain changelog entries for patch releases

I think the `[Unreleased]` patch release entry should be tracked on the release branch.

However, I think it is valuable to include published patch release entries on master after the release. That adds all release entries to one file for easy searching. This follows a similar pattern as the C driver NEWS file (e.g. both 1.26.0 and 1.25.4 are included on the master branch): https://github.com/mongodb/mongo-c-driver/blob/master/NEWS.

Reviewer: > I think it is valuable to include published patch release entries on master after the release.

That was unintentional/poor wording on my part. The _entries_ would still be present, but they would belong to a minor/major release section instead of a patch release section. There would be no patch release _sections_ in the main branch.

The response and changes are consistent with my understand and LGTM.

Reviewer: I would prefer that whatever approach we use also minimizes the amount of editing required during the release process (ideally, just removing `[Unreleased]`). If the entry must ultimately end up in both branches, I would prefer they be written to both branches _immediately_ rather than during the release process.

With that in mind, which of the following options do you prefer?

I'm personally in favor of Option A due to less likelihood for conflicts when cherry-picking from `master` to the release branch + minimal redundancy.

## Option A: patch release section in both master and release with implicit patch entries.

On `master`:
```md
# 3.11.0 [Unreleased]

* Entry targeting 3.11.0.

# 3.10.1 [Unreleased]

* Entry targeting 3.10.1. (Implicitly included in 3.11.0.)
```
On `releases/v3.10`:
```md
# 3.10.1 [Unreleased]

* Entry targeting 3.10.1.
```

## Option B: patch release section in both master and release with explicit patch entries.

On `master`:
```md
# 3.11.0 [Unreleased]

* Entry targeting 3.11.0.
* Entry targeting 3.10.1. (Explicitly included in 3.11.0.)

# 3.10.1 [Unreleased]

* Entry targeting 3.10.1.
```
On `releases/v3.10`:
```md
# 3.10.1 [Unreleased]

* Entry targeting 3.10.1.
```

## Option C: patch release section in release only.

On `master`:
```md
# 3.11.0 [Unreleased]

* Entry targeting 3.11.0.
* Entry targeting 3.10.1. (Explicitly included in 3.11.0.)
```
On `releases/v3.10`:
```md
# 3.10.1 [Unreleased]

* Entry targeting 3.10.1.
```

Author: I also prefer Option A. And the detailed description is much appreciated. The release process has been updated. CHANGELOG.md has been updated to include both upcoming patch (3.10.1) and minor (3.11.0) releases.

  `docs/`. This may require pushing the commit to a fork of the C++ Driver

>   repository and creating a pull request.

Reviewer: I propose removing the "Push the commit containing changes to `etc/` and `docs/`." sentence entirely and rewriting this instruction to direct user to simply create a pull request from a fork repo (as done with normal feature branches) due to branch protections on `master` forbidding direct pushes.

The "create a PR with post-release changes" can also be postponed to include addition of `[Unreleased]` stubs in the CHANGELOG.

## Update CHANGELOG.md for next release

CHANGELOG.md on the `master` branch contains sections for every release. This is intended to ease searching for changes among all releases.
CHANGELOG.md on a release branch (e.g. `releases/v1.2`) contains entries for all prior releases, and all patch releases on the `releases/v1.2` branch.

### For a minor release (e.g. r1.2.0):

Open a PR to the master branch to add a new empty `[Unreleased]` section to the CHANGELOG.md for the next minor release and next patch release. Example:

```md
## 1.3.0 [Unreleased]

<!-- Will contain entries for 1.3.0 release. -->

## 1.2.1 [Unreleased]

<!-- Will contain entries for 1.2.1 release. Entries are implicitly included in 1.3.0 release. -->

```

> Open a PR to the release branch (e.g. `releases/v1.2`) to add a new empty `[Unreleased]` section to the CHANGELOG.md for the next patch release. Example:

Reviewer: This step should be covered by the "Branch if necessary" section below that comes _after_ the release process has completed. Additionally, given relocation of the "Update CHANGELOG.md for next release" instructions, the required subsequent edit will likely involve _removal_ of the minor release section stub.

I recommend deferring the addition of `[Unreleased]` stubs (for both master and release branches) to _after_ the "Branch if necessary" section, but before docker/packaging instructions:

- Audit JIRA tickets.
- Remove `[Unreleased]` for current release _and commit_ (important).
- Create release tag.
- Run "make release" scripts.
- Add `[Unreleased]` stubs for next release(s) _and commit_ (important).
- Create a new release branch at this point if necessary.
- Push new/updated release branch to remote (note: PR not required!).
- Create PR for all post-release changes targeting `master`.

This allows for the release tag to be correctly timed _after_ removal of `[Unreleased]` in the CHANGELOG, remain stable post-release (no post-PR-merge recommit shenanigans), and cleanly workaround `master` branch protection by pushing the post-release commits to the _release_ branch first instead.

Preferably the post-release PR to `master` should use "Rebase and merge" to preserve the individual commits (a la cherry-picking), but a "Squash and merge" is also fine. Additionally, to avoid mixing/merging up the release branch history with `master`, the "Create PR for post-release rX.Y -> master" can be preceeded by a new (temporary, to-be-deleted post-PR) branch separate from `releases/vX.Y`.

Author: Updated changes include a `Update CHANGELOG.md post-release ...` after the `Branch if necessary` section with use of a `post-release-changes` branch.

    stable release branch.
  - Commit and push the symlink change:
    `git commit -am "Update symlink for r1.2.3"`
   - Switch back to the branch with documentation updates: `git checkout post-release-changes`.
- Wait a few minutes and verify mongocxx.org has updated.

## Update CHANGELOG.md post-release ...

CHANGELOG.md on the `master` branch contains sections for every release. This is intended to ease searching for changes among all releases.
CHANGELOG.md on a release branch (e.g. `releases/v1.2`) contains entries for patch releases of the minor version number tracked by the release branch (e.g. for 1.2.1, 1.2.2, 1.2.3, etc.), as well as all entries prior to the initial minor release (e.g. before 1.2.0).

### ... on the `master` branch

> Check out the `post-release-changes` branch created before.

Reviewer: ```suggestion
Check out the `post-release-changes` branch created before editing and generating documentation.
```
Clarify when the branch should have been created.

Ensure `[Unreleased]` is removed from the recently released section. Ensure the contents of the recently released section match the published release notes.

Ensure there are `[Unreleased]` sections for the next minor and patch releases. Example (if `1.2.3` was just released):

```md

> ## 1.3.0 [Unreleased]
> <!-- Will contain entries for the next minor release -->
> ## 1.2.4 [Unreleased]
> <!-- Will contain entries for the next patch release -->
> ## 1.2.3
> <!-- Contains published release notes -->

Reviewer: ```suggestion
## 1.3.0 [Unreleased]

<!-- Will contain entries for the next minor release -->

## 1.2.4 [Unreleased]

<!-- Will contain entries for the next patch release -->

## 1.2.3

<!-- Contains published release notes -->
```
Add whitespace.

```

Commit the change. Create a PR from the `post-release-changes` branch to merge to `master`.

> ### ... on the release branch

Reviewer: Suggest moving the "... on the release branch" step up so it comes before "... on the master branch". Tidying up the release branch first will chronologically group release-related commits together better than if `master` is updated first. (This is admittedly a very minor point + these two PRs will likely be done in parallel anyways, so this suggestion might be moot.)

Check out the release branch (e.g. `releases/v1.2`).

Update CHANGELOG.md to add an `[Unreleased]` section for the next patch release. Example (if `1.2.3` was just released):

```md
## 1.2.4 [Unreleased]
<!-- Will contain entries for the next patch release -->
## 1.2.3
<!-- Contains published release notes -->
```

> Commit and push this change to the release branch (no PR necessary for release branch).

Reviewer: Update this paragraph as necessary if we want to account for upcoming stable branch policies.

Author: I prefer to defer (elaborated here https://github.com/mongodb/mongo-cxx-driver/pull/1109/files#r1515187195)

    stable release branch.
  - Commit and push the symlink change:
    `git commit -am "Update symlink for r1.2.3"`
   - Switch back to the branch with documentation updates: `git checkout post-release-changes`.
- Wait a few minutes and verify mongocxx.org has updated.

## Update CHANGELOG.md post-release ...

CHANGELOG.md on the `master` branch contains sections for every release. This is intended to ease searching for changes among all releases.
CHANGELOG.md on a release branch (e.g. `releases/v1.2`) contains entries for patch releases of the minor version number tracked by the release branch (e.g. for 1.2.1, 1.2.2, 1.2.3, etc.), as well as all entries prior to the initial minor release (e.g. before 1.2.0).

> ### ... on the `master` branch
> 
> Check out the `post-release-changes` branch created before.
> 
> Ensure `[Unreleased]` is removed from the recently released section. Ensure the contents of the recently released section match the published release notes.
> 
> Ensure there are `[Unreleased]` sections for the next minor and patch releases. Example (if `1.2.3` was just released):
> 
> ```md
> ## 1.3.0 [Unreleased]
> <!-- Will contain entries for the next minor release -->
> ## 1.2.4 [Unreleased]
> <!-- Will contain entries for the next patch release -->
> ## 1.2.3
> <!-- Contains published release notes -->
> ```
> 
> Commit the change. Create a PR from the `post-release-changes` branch to merge to `master`.
> 

Reviewer: ```suggestion
```
(Nearly) all post-release _and_ pre-release changes that end up on the `master` branch will be the same as those on the release branch (regardless of minor or patch release). I believe it would be preferable to push all changes to the release branch (no intermediate PR required) before creating a PR to merge all release-related changes to the `master` branch.

Reviewer: Suggestion retracted.
